### PR TITLE
os/bluestore: fix no deferred writing for legacy OSDs

### PR DESCRIPTION
**Pull request by ifed01**
Fixes: https://tracker.ceph.com/issues/56488

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by aclamk — CHANGES_REQUESTED** [review]
I think "<=" commit should be dropped, and refactor commit amended to reflect logic change.

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -15639,31 +15638,22 @@ int BlueStore::_do_alloc_write(
...
-	prealloc_pos->length -= left;
-	prealloc_left -= left;
-	txc->statfs_delta.allocated() += left;
-	left = 0;
-	break;
       }
-    }
-    for (auto& p : extents) {
-      txc->allocated.insert(p.offset, p.length);
+      has_chunk2defer |= (l < prefer_deferred_size_snapshot);
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc]: @ifed01 ping... ?

**Comment by cfsnyder** [comment]
I agree with Adam's comments. This seems like a high priority fix since deferred writes are otherwise not happening at all for many clusters with default settings, right? It would be great to get this merged and backported to Pacific before the next Pacific release!

**Comment by ifed01** [comment]
> I agree with Adam's comments. This seems like a high priority fix since deferred writes are otherwise not happening at all for many clusters with default settings, right?
That's true. And I agree that Adam's comments are valid. Working on a better approach for now...

> It would be great to get this merged and backported to Pacific before the next Pacific release!

**Comment by cfsnyder** [comment]
@ifed01 would you mind sharing brief details about the better approach that you're working on? Do you think that this will be ready in time for the next Pacific release?

**Comment by ifed01** [comment]
> @ifed01 would you mind sharing brief details about the better approach that you're working on? Do you think that this will be ready in time for the next Pacific release?

I've just updated the PR with a modified implementation. This also adds some comments (and new test cases) to represent the ideas behind the patch. Please let me know if you have more questions

**Comment by ifed01** [comment]
@aclamk - ready for another review round

**Comment by aclamk** [comment]
@ifed01
Last week, I reviewed your solution and tested that it works.
From that point of view I think it can be merged.

However, I did not like a multitude of conditions that were needed and the fact that many unittest had to be adjusted to new revised `deferred count` changes.

I allowed myself to attempt different solution : #48490.
In any case, we need to decide on one, as they are incompatible.

More over - I believe we need a revamp of deferred write selection algorithm. 
The current one (unrelated decisions in _do_write_small, do_write_big, _do_allocate_write) make a fine-tuning impossible.

**Comment by ifed01** [comment]
Closing in favor of #48490

**Pull request closed without merging**